File: data structure and algorithms/Database/578.get-highest-answer-rate-question.java
/**
 * This LeetCode problem is a database problem, so the executable solution is
 * SQL rather than Java. The original Python file stores the SQL string; this
 * Java translation preserves that solution in a constant.
 *
 * Algorithm:
 * Group by question_id, compute answers divided by shows, and choose the
 * highest rate. Boolean expressions aggregate as 1/0 in MySQL.
 *
 * Database complexity:
 * One grouped scan over SurveyLog, usually improved by indexing question_id.
 */
class Solution {
    static final String SOLUTION_SQL = """
        SELECT question_id AS survey_log
        FROM SurveyLog
        GROUP BY question_id
        ORDER BY SUM(action = 'answer') / SUM(action = 'show') DESC, question_id
        LIMIT 1;
        """;
}

File: data structure and algorithms/Database/578.get-highest-answer-rate-question.go
package leetcode

// SolutionSQL578 is the SQL answer for LeetCode 578. The source repository
// stored this database problem in a Python file, so the Go translation keeps the
// executable SQL as a raw string constant instead of inventing unrelated Go code.
//
// Explanation: Aggregate answers divided by shows per question and order by rate, then question id for deterministic ties.
// Database complexity is dominated by grouping, sorting, and joins; indexes on
// the join/order/group columns are the relevant optimization.
const SolutionSQL578 = `SELECT question_id AS survey_log
FROM SurveyLog
GROUP BY question_id
ORDER BY SUM(action = 'answer') / SUM(action = 'show') DESC, question_id
LIMIT 1;`
